File: 1593. Split a String Into the Max Number of Unique Substrings/1593. Split a String Into the Max Number of Unique Substrings.go
// backtrack
func maxUniqueSplit(s string) int {
    if s == "" {
        return 0
    }
    dic := make(map[string]int)
    return backtrack(s,dic)
}
func backtrack(s string, dic map[string]int) int {
    ans := 0
    if s != "" {
        for i:=1;i<len(s)+1;i++{
            tmp := s[:i]
            _,ok := dic[tmp]
            if !ok{
                dic[tmp] = 1
                bt := 1+backtrack(s[i:],dic)
                if bt > ans{
                    ans = bt
                }
                delete(dic,tmp)
            }
        }
    }
    return ans
}

File: 1593. Split a String Into the Max Number of Unique Substrings/1593. Split a String Into the Max Number of Unique Substrings.py
# backtrack
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        if not s:
            return 0
        seen = set()
        return self.backtrack(seen,s)
    def backtrack(self,seen,s):
        ans = 0
        if s:
            for i in range(1,len(s)+1):
                tmp = s[:i]
                if tmp not in seen:
                    seen.add(tmp)
                    ans = max(ans,1+self.backtrack(seen,s[i:]))
                    seen.remove(tmp)
        return ans
